packages/next-upgrade/index.ts

```diff
@@ -0,0 +1,371 @@
+
+  if (!isWorkspace) return
+
+  console.log(
+    `${chalk.red('⚠️')} You seem to be in the root of a monorepo. ${chalk.blue('@next/upgrade')} should be run in a specific app directory within the monorepo.`
+  )
+
+  const response = await prompts(
+    {
+      type: 'confirm',
+      name: 'value',
+      message: 'Do you still want to continue?',
+      initial: false,
+    },
+    { onCancel: () => process.exit(0) }
+  )
+
+  if (!response.value) {
+    process.exit(0)
+  }
+}
+
+/*
+ * Logs the current version and returns the index of the current version specifier
+ * in the array ['canary', 'rc', 'latest']
+ */
+async function processCurrentVersion(showRc: boolean): Promise<number> {
+  const require = createRequire(import.meta.url)
+  const installedNextPackageJsonDir = require.resolve('next/package.json', {
+    paths: [process.cwd()],
+  })
+  const installedNextPackageJson = JSON.parse(
+    fs.readFileSync(installedNextPackageJsonDir, 'utf8')
+  )
+  let installedNextVersion = installedNextPackageJson.version
+
+  if (installedNextVersion == null) {
+    return 0
+  }
+
+  console.log(
+    `You are currently using ${chalk.blue('Next.js ' + installedNextVersion)}`
+  )
+
+  if (installedNextVersion.includes('canary')) {
+    return 0
+  }
+  if (installedNextVersion.includes('rc')) {
+    return 1 // If rc is not available, will default to latest's index
+  }
+  return showRc ? 2 : 1 // "latest" is 1 or 2 depending on if rc is shown as an option
+}
+
+async function getPackageManager(_packageJson: any): Promise<PackageManager> {
+  const packageManagers = {
+    pnpm: 'pnpm-lock.yaml',
+    yarn: 'yarn.lock',
+    npm: 'package-lock.json',
+    bun: 'bun.lockb',
+  }
+
+  function findLockFile(dir: string): PackageManager[] {
+    const packageJsonPath = path.join(dir, 'package.json')
+    if (fs.existsSync(packageJsonPath)) {
+      let detectedPackageManagers: PackageManager[] = []
+      let packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
+      if (packageJson.packageManager) {
+        // corepack
+        let packageManagerName = packageJson.packageManager.split(
+          '@'
+        )[0] as PackageManager
+        if (packageManagerName in packageManagers) {
+          return [packageManagerName]
+        }
+      }
+      for (const [packageManager, lockFile] of Object.entries(
+        packageManagers
+      )) {
+        const lockFilePath = path.join(dir, lockFile)
+        if (fs.existsSync(lockFilePath)) {
+          detectedPackageManagers.push(packageManager as PackageManager)
+        }
+      }
+      if (detectedPackageManagers.length !== 0) {
+        return detectedPackageManagers
+      }
+    }
+    const parentDir = path.dirname(dir)
+    if (parentDir !== dir) {
+      return findLockFile(parentDir)
+    }
+    return []
+  }
+
+  let realPath = fs.realpathSync(process.cwd())
+  const detectedPackageManagers = findLockFile(realPath)
```

Comment:
If it's there, we should probably prefer the top-level `packageManager` field in package.json above the rest of this logic.

Reply:
Oh, just saw that this does happen in `findLockFile` 👍 Could lift it out of there just to avoid confusion with the name.